storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	// Update the peer in the swarm.
+	encoded_ih_L := IPver + "_L_" + ih.String()
+	ct := ps.getClock()
+
+	_, err := conn.Do("HSET", encoded_ih_L, pk, ct)
+
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.Do("HSET", IPver, encoded_ih_L, ct)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) DeleteLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: DeleteLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encoded_ih_L := IPver + "_L_" + ih.String()
+
+	DelNum, err := conn.Do("HDEL", encoded_ih_L, pk)
+	if err != nil {
+		return err
+	}
+	if DelNum.(int64) == 0 {
+		return storage.ErrResourceDoesNotExist
+	}
+
+	return nil
+}
+
+func (ps *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: GraduateLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih
+	encoded_ih_S := IPver + "_S_" + encoded_ih
+
+	_, err := conn.Do("HDEL", encoded_ih_L, pk)
+	if err != nil {
+		return err
+	}
+
+	// Update the peer in the swarm.
+	ct := ps.getClock()
+	_, err = conn.Do("HSET", encoded_ih_S, pk, ct)
+
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.Do("HSET", IPver, encoded_ih_S, ct)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	IPver := announcer.IP.AddressFamily.String()
+	log.Debug("storage: AnnouncePeers", log.Fields{
+		"InfoHash": ih.String(),
+		"seeder":   seeder,
+		"numWant":  numWant,
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", announcer.ID.String(), announcer.IP.String(), IPver, announcer.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers, err := conn.Do("HKEYS", encoded_ih_L)
```

Comment:
Hypothetically, for a swarm with 50,000 leechers, will this actually retrieve all 50,000 peer keys?

Reply:
It would, perhaps some tricks with HSCAN could work here.

Reply:
@mrd0ll4r yes, redis [HKEYS](https://redis.io/commands/hkeys) will return all relevant leechers(redis hash `field`) of the `HashInfo`(redis hash `key`)
@pushrax you are right, maybe we can use `HSCAN` get Peers in batches after all redis is single thread designed.